include/albatross/src/models/sparse_gp.hpp

```diff
@@ -315,23 +321,100 @@ class SparseGaussianProcessRegression
     RtR = RtR.transpose() * RtR;
     const Eigen::MatrixXd B = Eigen::MatrixXd::Identity(m, m) + RtR;
 
-    const auto B_ldlt = B.ldlt();
-    const Eigen::MatrixXd L_uu_inv =
-        K_uu_llt.matrixL().solve(Eigen::MatrixXd::Identity(m, m));
-    const Eigen::MatrixXd BiLi = B_ldlt.solve(L_uu_inv);
-    const Eigen::MatrixXd RtRBiLi = RtR * B_ldlt.solve(L_uu_inv);
-    const auto LT = K_uu_llt.matrixL().transpose();
+    const Eigen::SerializableLDLT B_ldlt(B);
 
-    Eigen::VectorXd v = BiLi.transpose() * P * A_llt.solve(targets.mean);
+    SparseGPComponents components = {
+        K_uu_llt, P, A_llt, RtR, B_ldlt, targets.mean,
+    };
 
-    const Eigen::MatrixXd C_inv = LT.solve(RtRBiLi);
+    return components;
+  }
 
+  template <typename FeatureType>
+  auto _fit_impl(const std::vector<FeatureType> &features,
+                 const MarginalDistribution &targets) const {
+
+    // Determine the set of inducing points, u.
+    const auto u =
+        inducing_point_strategy_(this->covariance_function_, features);
+
+    const auto sc = sparse_components(features, u, targets);
+
+    const Eigen::Index m = sc.K_uu_llt.rows();
+    const Eigen::MatrixXd L_uu_inv =
+        sc.K_uu_llt.matrixL().solve(Eigen::MatrixXd::Identity(m, m));
+    const Eigen::MatrixXd BiLi = sc.B_ldlt.solve(L_uu_inv);
+    const Eigen::MatrixXd RtRBiLi = sc.RtR * sc.B_ldlt.solve(L_uu_inv);
+    const auto LT = sc.K_uu_llt.matrixL().transpose();
+    Eigen::VectorXd v = BiLi.transpose() * sc.P * sc.A_llt.solve(sc.y);
+
+    const Eigen::MatrixXd C_inv = LT.solve(RtRBiLi);
     DirectInverse solver(C_inv);
 
     using InducingPointFeatureType = typename std::decay<decltype(u[0])>::type;
     return Fit<GPFit<DirectInverse, InducingPointFeatureType>>(u, solver, v);
   }
 
+  template <typename FeatureType>
+  double log_likelihood(const RegressionDataset<FeatureType> &dataset) const {
+
+    const auto u =
+        inducing_point_strategy_(this->covariance_function_, dataset.features);
+
+    const SparseGPComponents sc =
+        sparse_components(dataset.features, u, dataset.targets);
+
+    // The log likelihood for y ~ N(0, K) is:
+    //
+    //   L = 1/2 (n log(2 pi) + log(|K|) + y^T K^-1 y)
+    //
+    // where in our case we have K = A + Q_ff
+
+    // First we get the determinant, |K|:
+    //
+    //   |K| = |A + Q_ff|
+    //       = |K_uu + K_uf A^-1 K_fu| |K_uu^-1| |A|
+    //       = |LL^T + K_uf A^-1 K_fu| |L^-T L^-1| |A|
+    //       = |L^-1| |LL^T + K_uf A^-1 K_fu| |L^-T| |A|
+    //       = |I + L^-1 K_uf A^-1 K_fu L^-T| |A|
+    //       = |I + P A^-1 P^T| |A|
+    //       = |B| |A|
+    //
+    // which is derived here (though there are errors in their derivation):
+    //   https://bwengals.github.io/pymc3-fitcvfe-implementation-notes.html
+
+    const double log_det_a = sc.A_llt.log_determinant();
+    const double log_det_b = sc.B_ldlt.vectorD().array().log().sum();
```

Comment:
Why does `B`s determinant not involve the `L` part of the decomposition?

Reply:
That's because the `LDLT` imposes that the `L` has a unit diagonal.  Maybe I'll break this out into a little helper function.